File: backend/src/main/java/com/propertyapp/model/Property.java
package com.propertyapp.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "Properties")
public class Property {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "property_id")
    private Long propertyId;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    private User seller;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "property_type", nullable = false)
    private PropertyType propertyType;
    
    @NotBlank(message = "Title is required")
    @Size(max = 200, message = "Title must not exceed 200 characters")
    @Column(name = "title", nullable = false)
    private String title;
    
    @Column(name = "description", columnDefinition = "NVARCHAR(MAX)")
    private String description;
    
    @Column(name = "property_size", precision = 10, scale = 2)
    private BigDecimal propertySize;
    
    @NotNull(message = "Price is required")
    @DecimalMin(value = "0.0", inclusive = false, message = "Price must be greater than 0")
    @Column(name = "price", precision = 15, scale = 2, nullable = false)
    private BigDecimal price;
    
    @NotBlank(message = "Address is required")
    @Size(max = 500, message = "Address must not exceed 500 characters")
    @Column(name = "address", nullable = false)
    private String address;
    
    @NotBlank(message = "City is required")
    @Size(max = 100, message = "City must not exceed 100 characters")
    @Column(name = "city", nullable = false)
    private String city;
    
    @NotBlank(message = "State is required")
    @Size(max = 100, message = "State must not exceed 100 characters")
    @Column(name = "state", nullable = false)
    private String state;
    
    @NotBlank(message = "Pincode is required")
    @Size(max = 10, message = "Pincode must not exceed 10 characters")
    @Column(name = "pincode", nullable = false)
    private String pincode;
    
    @Column(name = "latitude", precision = 10, scale = 8)
    private BigDecimal latitude;
    
    @Column(name = "longitude", precision = 11, scale = 8)
    private BigDecimal longitude;
    
    @Column(name = "property_images", columnDefinition = "NVARCHAR(MAX)")
    private String propertyImages; // JSON array of image URLs
    
    @Size(max = 500, message = "PTR document path must not exceed 500 characters")
    @Column(name = "ptr_document")
    private String ptrDocument;
    
    @Column(name = "is_active")
    private Boolean isActive = true;
    
    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    // Relationships
    @JsonIgnore
    @OneToMany(mappedBy = "property", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Inquiry> inquiries;
    
    // Constructors
    public Property() {}
    
    public Property(User seller, PropertyType propertyType, String title, 
                   BigDecimal price, String address, String city, String state, String pincode) {
        this.seller = seller;
        this.propertyType = propertyType;
        this.title = title;
        this.price = price;
        this.address = address;
        this.city = city;
        this.state = state;
        this.pincode = pincode;
    }
    
    // Getters and Setters
    public Long getPropertyId() { return propertyId; }
    public void setPropertyId(Long propertyId) { this.propertyId = propertyId; }
    
    public User getSeller() { return seller; }
    public void setSeller(User seller) { this.seller = seller; }
    
    public PropertyType getPropertyType() { return propertyType; }
    public void setPropertyType(PropertyType propertyType) { this.propertyType = propertyType; }
    
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public BigDecimal getPropertySize() { return propertySize; }
    public void setPropertySize(BigDecimal propertySize) { this.propertySize = propertySize; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) { this.price = price; }
    
    public String getAddress() { return address; }
    public void setAddress(String address) { this.address = address; }
    
    public String getCity() { return city; }
    public void setCity(String city) { this.city = city; }
    
    public String getState() { return state; }
    public void setState(String state) { this.state = state; }
    
    public String getPincode() { return pincode; }
    public void setPincode(String pincode) { this.pincode = pincode; }
    
    public BigDecimal getLatitude() { return latitude; }
    public void setLatitude(BigDecimal latitude) { this.latitude = latitude; }
    
    public BigDecimal getLongitude() { return longitude; }
    public void setLongitude(BigDecimal longitude) { this.longitude = longitude; }
    
    public String getPropertyImages() { return propertyImages; }
    public void setPropertyImages(String propertyImages) { this.propertyImages = propertyImages; }
    
    public String getPtrDocument() { return ptrDocument; }
    public void setPtrDocument(String ptrDocument) { this.ptrDocument = ptrDocument; }
    
    public Boolean getIsActive() { return isActive; }
    public void setIsActive(Boolean isActive) { this.isActive = isActive; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    
    public List<Inquiry> getInquiries() { return inquiries; }
    public void setInquiries(List<Inquiry> inquiries) { this.inquiries = inquiries; }
}

File: ios-app/PropertyListing/Models/Property.swift
import Foundation

// MARK: - Property Model
struct Property: Codable {
    let id: Int?
    let propertyType: PropertyType
    let title: String
    let description: String?
    let propertySize: Double
    let price: Double
    let address: String
    let city: String
    let state: String
    let pincode: String
    let latitude: Double?
    let longitude: Double?
    let propertyImages: String?
    let ptrDocument: String?
    let seller: User?
    let isActive: Bool?
    let createdAt: String?
    let updatedAt: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case propertyType = "property_type"
        case title
        case description
        case propertySize = "property_size"
        case price
        case address
        case city
        case state
        case pincode
        case latitude
        case longitude
        case propertyImages = "property_images"
        case ptrDocument = "ptr_document"
        case seller
        case isActive = "is_active"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - Property Type Enum
enum PropertyType: String, Codable, CaseIterable {
    case land = "LAND"
    case flat = "FLAT"
    case rowHouse = "ROW_HOUSE"
    case bungalow = "BUNGALOW"
    
    var displayName: String {
        switch self {
        case .land:
            return "Land"
        case .flat:
            return "Flat"
        case .rowHouse:
            return "Row House"
        case .bungalow:
            return "Bungalow"
        }
    }
    
    var icon: String {
        switch self {
        case .land:
            return "map"
        case .flat:
            return "building"
        case .rowHouse:
            return "house"
        case .bungalow:
            return "house.fill"
        }
    }
}

// MARK: - Property Request Model
struct PropertyRequest: Codable {
    let propertyType: PropertyType
    let title: String
    let description: String
    let propertySize: Double
    let price: Double
    let address: String
    let city: String
    let state: String
    let pincode: String
    let latitude: Double?
    let longitude: Double?
    let propertyImages: String?
    let ptrDocument: String?
    
    enum CodingKeys: String, CodingKey {
        case propertyType = "property_type"
        case title
        case description
        case propertySize = "property_size"
        case price
        case address
        case city
        case state
        case pincode
        case latitude
        case longitude
        case propertyImages = "property_images"
        case ptrDocument = "ptr_document"
    }
}
